src/main/java/org/dbos/apiary/xa/XAConnection.java

```diff
@@ -8,30 +8,89 @@
+    public XAConnection(XADBConnection postgresConnection, XADBConnection mysqlConnection) throws SQLException {
+        assert(postgresConnection != null);
+        assert(mysqlConnection != null);
+        this.postgresConnection = postgresConnection;
+        this.mysqlConnection = mysqlConnection;
         // TODO: implement building connections to databases.
         // Please refer to the implementation of PostgresConnection.
     }
 
     @Override
     public FunctionOutput callFunction(String functionName, WorkerContext workerContext, String service, long execID, long functionID, Object... inputs) throws Exception {
         // TODO: may pass more parameters to XAContext.
-        XAContext ctxt = new XAContext(workerContext, service, execID, functionID);
+        XAContext ctxt = new XAContext(this, workerContext, service, execID, functionID);
+        ApiaryXID xid  = ApiaryXID.fromLong(xidCounter.getAndIncrement());
         FunctionOutput f = null;
+        boolean committed = false;
+        boolean ended = false;
+        // TODO: transaction manager does 2PC here to commit or abort the transaction.
+
         try {
+            // Start XA transaction in underlying databases
+            getXAConnection(PostgresDBType).XAStart(xid);
+            getXAConnection(MySQLDBType).XAStart(xid);
             // The function would contain transactions across multiple databases.
             f = workerContext.getFunction(functionName).apiaryRunFunction(ctxt, inputs);
+            // End XA transaction in underlying databases
+            getXAConnection(PostgresDBType).XAEnd(xid);
+            getXAConnection(MySQLDBType).XAEnd(xid);
+            ended =true;
+
+            // Prepare-phase
+            if (getXAConnection(PostgresDBType).XAPrepare(xid) && getXAConnection(MySQLDBType).XAPrepare(xid)) {
+                // TODO: persist commit decision ?
+                // Commit-phase
+                getXAConnection(PostgresDBType).XACommit(xid);
+                getXAConnection(MySQLDBType).XACommit(xid);
+                committed = true;
+            }
         } catch (Exception e) {
-            e.printStackTrace();
+            //e.printStackTrace();
+        }
+
+        if (committed == false) {
+            if (ended == false) {
+                getXAConnection(PostgresDBType).XAEnd(xid);
+                getXAConnection(MySQLDBType).XAEnd(xid);
+            }
+            // TODO: persist abort decision ?
+            // Rollback XA transaction in underlying databases
+            getXAConnection(PostgresDBType).XARollback(xid);
+            getXAConnection(MySQLDBType).XARollback(xid);
+        }
+        if (f == null) {
+            // Hack. Need to return a non-null function output in case of Exceptions (e.g., concurrency control conflicts from PG) thrown from underlying databases.
```

Comment:
If a concurrency control conflict happens, we might want to retry the execution. Then return null if a non-concurrency control conflict exception happens, which cannot be resolved so we will surface the exception. For example, the apiary code: https://github.com/DBOS-project/apiary/blob/main/src/main/java/org/dbos/apiary/postgres/PostgresConnection.java#L139

Reply:
Fixed